### PR TITLE
Spyglass log viewer: handle bash formatting control sequences

### PR DESCRIPTION
**Pull request by ibzib**
Applies [Gubernator's code](https://github.com/kubernetes/test-infra/blob/931beb7a716ae9a7a93729c225396d4fe5f9389b/gubernator/static/build.js#L131-L159) for handling bash formatting control sequences to the Spyglass build log viewer.

/assign @Katharine @BenTheElder 
screenshot from 2018-11-12 14-21-19

### CODE REVIEW

File: prow/spyglass/viewers/buildlog/buildlog_template.go [hunk 1]
```diff
@@ -41,6 +41,7 @@ var buildLogTemplateText = `<style>
 .skipped {
   display: none;
 }
+.shown {}
```
Comment (Katharine) [inline prow/spyglass/viewers/buildlog/buildlog_template.go]: Is this supposed to have some content? It's used, but empty.

Comment (ibzib) [reply prow/spyglass/viewers/buildlog/buildlog_template.go]: I just left it in as a placeholder. Not really necessary though

Comment (BenTheElder) [reply prow/spyglass/viewers/buildlog/buildlog_template.go]: imo you should remove that then

**Comment by Katharine** [comment]
Actually...
/lgtm cancel

(But I am probably too late now.)

File: prow/cmd/deck/static/spyglass/spyglass.ts [hunk 2]
```diff
@@ -51,7 +51,17 @@ function requestReload(name: string, body: string): void {
 // Insert rendered view into page and remove loading wheel
 function insertView(name: string, content: string): void {
   document.getElementById(name + "-loading")!.style.display = "none";
-  document.getElementById(name + "-view")!.innerHTML = content;
+
+  const view = document.getElementById(name + "-view");
```
Comment (Katharine) [inline prow/cmd/deck/static/spyglass/spyglass.ts]: Put the `!` here so you don't need it everywhere else you use `view`.

```suggestion
  const view = document.getElementById(name + "-view")!;
```

File: prow/cmd/deck/static/spyglass/spyglass.ts [hunk 3]
```diff
@@ -51,7 +51,17 @@ function requestReload(name: string, body: string): void {
...
+  view!.innerHTML = content;
+  // This is an icky workaround until we have viewer-specific scripts
+  // https://github.com/kubernetes/test-infra/issues/8967
+  if (name === "build-log-viewer") {
+    const shown = view!.getElementsByClassName("shown");
```
Comment (Katharine) [inline prow/cmd/deck/static/spyglass/spyglass.ts]:
```suggestion
    const shown = view.getElementsByClassName("shown");
```

File: prow/cmd/deck/static/spyglass/spyglass.ts [hunk 4]
```diff
@@ -51,7 +51,17 @@ function requestReload(name: string, body: string): void {
 // Insert rendered view into page and remove loading wheel
 function insertView(name: string, content: string): void {
   document.getElementById(name + "-loading")!.style.display = "none";
-  document.getElementById(name + "-view")!.innerHTML = content;
+
+  const view = document.getElementById(name + "-view");
+  view!.innerHTML = content;
```
Comment (Katharine) [inline prow/cmd/deck/static/spyglass/spyglass.ts]:
```suggestion
  view.innerHTML = content;
```

File: prow/cmd/deck/static/spyglass/spyglass.ts [hunk 5]
```diff
@@ -61,13 +71,18 @@ function insertLoading(name: string): void {
 
 window.onload = loadViews;
 
+function showElem(elem: HTMLElement) {
```
Comment (Katharine) [inline prow/cmd/deck/static/spyglass/spyglass.ts]: :robot: `Avoid implicit any by always specifying return types` (this is supposed to be a compile error...)

```suggestion
function showElem(elem: HTMLElement): void {
```

File: prow/cmd/deck/static/spyglass/spyglass.ts [hunk 6]
```diff
@@ -99,6 +113,39 @@ function toggleExpansion(bodyId: string, expanderId: string): void {
   }
 }
 
+// given a string containing ansi formatting directives, return a new one
+// with designated regions of text marked with the appropriate color directives,
+// and with all unknown directives stripped
+function ansiToHTML(orig: string) {
```
Comment (Katharine) [inline prow/cmd/deck/static/spyglass/spyglass.ts]: :robot: `Avoid implicit any by always specifying return types` 

```suggestion
function ansiToHTML(orig: string): string {
```

File: prow/cmd/deck/static/spyglass/spyglass.ts [hunk 7]
```diff
@@ -99,6 +113,39 @@ function toggleExpansion(bodyId: string, expanderId: string): void {
...
+  // Given a cmd (like "32" or "0;97"), some enclosed body text, and the original string,
+  // either return the body wrapped in an element to achieve the desired result, or the
+  // original string if nothing works.
+  function annotate(cmd: string, body: string, orig: string) {
```
Comment (Katharine) [inline prow/cmd/deck/static/spyglass/spyglass.ts]: :robot: `Avoid implicit any by always specifying return types` 

```suggestion
  function annotate(cmd: string, body: string, orig: string): string {
```

File: prow/cmd/deck/static/spyglass/spyglass.ts [hunk 8]
```diff
@@ -99,6 +113,39 @@ function toggleExpansion(bodyId: string, expanderId: string): void {
...
+      return '<em>' + body + '</em>';
+    else if (30 <= code && code <= 37) // foreground color
+      return '<span class="ansi-' + (code - 30) + '">' + body + '</span>';
+    else if (90 <= code && code <= 97) // foreground color, bright
+      return '<span class="ansi-' + (code - 90 + 8) + '">' + body + '</span>';
+    return body;  // fallback: don't change anything
+  }
+  // Find commands, optionally followed by a bold command, with some content, then a reset command.
+  // Unpaired commands are *not* handled here, but they're very uncommon.
+  var filtered = orig.replace(/\033\[([0-9;]*)\w(\033\[1m)?([^\033]*?)\033\[0m/g, function(match: string, cmd: string, bold: string, body: string, offset: number, str: string) {
```
Comment (Katharine) [inline prow/cmd/deck/static/spyglass/spyglass.ts]: :robot: `Prefer const where the value is never reassigned.`
:robot: `Prefer arrow functions.`

```suggestion
  const filtered = orig.replace(/\033\[([0-9;]*)\w(\033\[1m)?([^\033]*?)\033\[0m/g, (match: string, cmd: string, bold: string, body: string, offset: number, str: string) => {
```

File: prow/cmd/deck/static/spyglass/spyglass.ts [hunk 9]
```diff
@@ -99,6 +113,39 @@ function toggleExpansion(bodyId: string, expanderId: string): void {
...
+    if (bold !== undefined)  // normal code + bold
+      return '<em>' + annotate(cmd, body, str) + '</em>';
+    return annotate(cmd, body, str);
+  });
+  // Strip out anything left over.
+  return filtered.replace(/\033\[([0-9;]*\w)/g, function(match: string, cmd: string, offset: number, str: string) {
```
Comment (Katharine) [inline prow/cmd/deck/static/spyglass/spyglass.ts]: :robot: `Prefer arrow functions.`

```suggestion
  return filtered.replace(/\033\[([0-9;]*\w)/g, (match: string, cmd: string, offset: number, str: string) => {
```

(Callback types can usually be inferred, but in this case the type definitions on `String.prototype.replace` are too vague. You can however omit the return type and, technically, the type of `match`.)

File: prow/cmd/deck/static/spyglass/spyglass.ts [hunk 10]
```diff
@@ -99,6 +113,39 @@ function toggleExpansion(bodyId: string, expanderId: string): void {
...
 
+// given a string containing ansi formatting directives, return a new one
+// with designated regions of text marked with the appropriate color directives,
+// and with all unknown directives stripped
+function ansiToHTML(orig: string) {
+  // Given a cmd (like "32" or "0;97"), some enclosed body text, and the original string,
+  // either return the body wrapped in an element to achieve the desired result, or the
+  // original string if nothing works.
+  function annotate(cmd: string, body: string, orig: string) {
+    var code = +(cmd.replace('0;', ''));
```
Comment (Katharine) [inline prow/cmd/deck/static/spyglass/spyglass.ts]: :robot: `Prefer const where the value is never reassigned.`

```suggestion
    const code = +(cmd.replace('0;', ''));
```

File: prow/cmd/deck/static/spyglass/spyglass.ts [hunk 11]
```diff
@@ -99,6 +113,39 @@ function toggleExpansion(bodyId: string, expanderId: string): void {
...
+    if (code === 0) // reset
```
Comment (Katharine) [inline prow/cmd/deck/static/spyglass/spyglass.ts]: :robot: `Always use braces.`

File: prow/cmd/deck/static/spyglass/spyglass.ts [hunk 12]
```diff
@@ -99,6 +113,39 @@ function toggleExpansion(bodyId: string, expanderId: string): void {
...
+    else if (30 <= code && code <= 37) // foreground color
+      return '<span class="ansi-' + (code - 30) + '">' + body + '</span>';
+    else if (90 <= code && code <= 97) // foreground color, bright
+      return '<span class="ansi-' + (code - 90 + 8) + '">' + body + '</span>';
+    return body;  // fallback: don't change anything
+  }
+  // Find commands, optionally followed by a bold command, with some content, then a reset command.
+  // Unpaired commands are *not* handled here, but they're very uncommon.
+  var filtered = orig.replace(/\033\[([0-9;]*)\w(\033\[1m)?([^\033]*?)\033\[0m/g, function(match: string, cmd: string, bold: string, body: string, offset: number, str: string) {
+    if (bold !== undefined)  // normal code + bold
```
Comment (Katharine) [inline prow/cmd/deck/static/spyglass/spyglass.ts]: :robot: `Always use braces.`

**Comment by Katharine** [comment]
/lgtm
/hold

Feel free to unhold provided this still works after my suggestions.

**Pull request merged**